src/client/Microsoft.Identity.Client/Internal/ClientCredential/ClientAssertionDelegateCredential.cs

```diff
@@ -0,0 +1,109 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Security.Cryptography.X509Certificates;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Identity.Client.Internal.Requests;
+using Microsoft.Identity.Client.OAuth2;
+using Microsoft.Identity.Client.PlatformsCommon.Interfaces;
+using Microsoft.Identity.Client.TelemetryCore;
+
+namespace Microsoft.Identity.Client.Internal.ClientCredential
+{
+    /// <summary>
+    /// Handles client assertions supplied via a delegate that returns
+    /// <see cref="AssertionResponse"/> (JWT + optional certificate).
+    /// </summary>
+    internal sealed class ClientAssertionDelegateCredential : IClientCredential
+    {
+        private readonly Func<AssertionRequestOptions, CancellationToken, Task<AssertionResponse>> _assertionDelegate;
+
+        public ClientAssertionDelegateCredential(
+            Func<AssertionRequestOptions, CancellationToken, Task<AssertionResponse>> assertionDelegate)
+        {
+            _assertionDelegate = assertionDelegate ?? throw new ArgumentNullException(nameof(assertionDelegate));
+        }
+
+        public AssertionType AssertionType => AssertionType.ClientAssertion;
+
+        private X509Certificate2 _cachedCertificate;
+
+        internal X509Certificate2 PeekCertificate(string clientId)
+        {
+            // Return the cached value if we already probed once
+            if (_cachedCertificate != null)
+            {
+                return _cachedCertificate;
+            }
+
+            try
+            {
+                var probeOpts = new AssertionRequestOptions
+                {
+                    ClientID = clientId,
+                };
+
+                var resp = _assertionDelegate(probeOpts, CancellationToken.None)
+                                 .ConfigureAwait(false)
+                                 .GetAwaiter()
+                                 .GetResult();
+
+                _cachedCertificate = resp?.TokenBindingCertificate;
+            }
+            catch
```

Comment:
Is this empty catch needed?

Reply:
Does it make sense to acpture the exception and throw if the invcation of the delegate fails?